### PR TITLE
Dev/run single reactant

### PR DESCRIPTION
**Pull request by bp-kelley**
Adds the ability to run a single reactant at a time through a reaction (runReactant versus runReactants)
Adds reduceProductToSidechain that takes the product from a reaction, and reduces the product to the sidechains added by the reaction.

This enables a few operations, such as identifying and extracting sidechains created by a particular reagent.

### CODE REVIEW

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp [hunk 1]
```diff
@@ -116,7 +132,7 @@ void recurseOverReactantCombinations(
     VectVectMatchVectType &matchesPerProduct, unsigned int level,
     VectMatchVectType combination) {
   unsigned int nReactants = matchesByReactant.size();
-  URANGE_CHECK(level, nReactants - 1);
+  RANGE_CHECK(0, level, nReactants - 1);
```
Comment (greglandrum) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: should use URANGE_CHECK

File: Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py [hunk 2]
```diff
@@ -521,5 +521,85 @@ def test21CheckRawIters(self):
...
+    # from
+    # A Collection of Robust Organic Synthesis Reactions for In Silico Molecule Design
+    # Markus Hartenfeller,*, Martin Eberle, Peter Meier, Cristina Nieto-Oberhuber,
+    # Karl-Heinz Altmann, Gisbert Schneider, Edgar Jacoby, and Steffen Renner
+    # Novartis Institutes for BioMedical Research, Novartis Pharma AG, Forum 1,
+    # Novartis Campus, CH-4056 Basel, Switzerland Swiss Federal Institute of Technology (ETH)
+    #  Zurich, Switzerland
+    smirks_thiourea = "[N;$(N-[#6]):3]=[C;$(C=S):1].[N;$(N[#6]);!$(N=*);!$([N-]);!$(N#*);!$([ND3]);!$([ND4]);!$(N[O,N]);!$(N[C,S]=[S,O,N]):2]>>[N:3]-[C:1]-[N+0:2]"
+    rxn = rdChemReactions.ReactionFromSmarts(smirks_thiourea)
+    reagents = map(Chem.MolFromSmiles, ['C=CCN=C=S', 'NCc1ncc(Cl)cc1Br'])
```
Comment (greglandrum) [inline Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py]: Better to use a list comprehension here: `reagents = [Chem.MolFromSmiles(x) for x in ('C=CCN=C=S', 'NCc1ncc(Cl)cc1Br')]`.
If you're going to stick with `map`, you need to explicitly convert it to either a list or a tuple in order for this code to work with Python3

Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py]: I keep forgetting that Guido dislikes map, but the muscle memory is strong with this one.

**Comment by bp-kelley** [comment]
Is there anything else you would like to see for this feature?  swig wrap perhaps?

**Comment by greglandrum** [comment]
Hmm, did you not see the request for more information on Slack?

-Greg

**Comment by bp-kelley** [comment]
No, probably operator error here, I'll check my settings on slack.

**Comment by greglandrum** [comment]
This behavior isn't really nice:
```In [3]: rxn = AllChem.ReactionFromSmarts('[C:1][N:2].[CH:3]=[O:4]>>[C:1][N:2]C1CCC1O[C:3]=[O:4]')

In [4]: ps = rxn.RunReactants((Chem.MolFromSmiles('CCN'),Chem.MolFromSmiles('CC(=O)')))

In [5]: p = ps[0][0]

In [6]: fp = Chem.MolFromSmiles('CCNC1CCC1OC(C)=O')

In [7]: Chem.MolToSmiles(p,True)
Out[7]: 'CCNC1CCC1OC(C)=O'

In [8]: Chem.MolToSmiles(AllChem.ReduceProductToSideChains(p),True)
Out[8]: 'C[1*:1].C[3*:3]'

In [9]: Chem.MolToSmiles(AllChem.ReduceProductToSideChains(fp),True)
---------------------------------------------------------------------------
ArgumentError                             Traceback (most recent call last)
<ipython-input-9-6e8a8543d067> in <module>()
----> 1 Chem.MolToSmiles(AllChem.ReduceProductToSideChains(fp),True)

ArgumentError: Python argument types in
    rdkit.Chem.rdChemReactions.ReduceProductToSideChains(Mol)
did not match C++ signature:
    ReduceProductToSideChains(boost::shared_ptr<RDKit::ROMol> {lvalue} product, bool addDummyAtoms=True)
```

It's solvable by making the `product` argument to `reduceProductToSideChains()` to a `const ROMOL_SPTR &`

**Comment by greglandrum** [comment]
Thought it may be worth thinking about adding a version of `reduceProductToSideChains()` that can take an arbitrary molecule (that could be a product) and the reaction and do this work.

**Comment by bp-kelley** [comment]
The arbitrary molecule reduction seems like work!

**Comment by greglandrum** [comment]
Well, yeah. That's what I meant by "may be worth thinking about" :-)

**Comment by greglandrum** [comment]
@bp-kelley,  just to refresh here, this is waiting on the minor fix for this comment (https://github.com/rdkit/rdkit/pull/705#issuecomment-166516322) and then it's ready to go.

**Comment by bp-kelley** [comment]
@greglandrum I got a bit sidetracked seeing how a python held ROMol* could properly be converted to a const shared_ptr<ROMol> and be deleted properly :)  There's a LOT of magic going on:

http://www.boost.org/doc/libs/1_43_0/boost/python/converter/shared_ptr_from_python.hpp

mostly in the shared_ptr_deleter which combines shared_ptr semantics with python refcounting semantics...

**Comment by bp-kelley** [comment]
@greglandrum - I think I can reimplement this using the same technique as isMolProductOfReaction, but not today :)

File: Code/GraphMol/ChemReactions/Reaction.h [hunk 3]
```diff
@@ -198,6 +198,15 @@ class ChemicalReaction {
   */
   std::vector<MOL_SPTR_VECT> runReactants(const MOL_SPTR_VECT reactants) const;
 
+  //! Runs a single reactant against a single reactant template
+  /*!
+     \param reactant The single reactant to use
+
+     \param reactantTemplateIdx the reactant template to target in the reaction
```
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/Reaction.h]: Needs docstring on what the return'd std::vector<MOL_SPTR_VECT> is

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp [hunk 4]
```diff
@@ -63,50 +77,52 @@ bool getReactantMatches(const MOL_SPTR_VECT &reactants,
...
+      //   Yes, in this case there are only 2 unique products, but that's
+      //   a factor of the reactants' symmetry.
+      //
+      //   There's no particularly straightforward way of solving this problem
+      //   of recognizing cases
+      //   where we should give all matches and cases where we shouldn't; it's
+      //   safer to just
+      //   produce everything and let the client deal with uniquifying their
+      //   results.
+      int matchCount = SubstructMatch(*(reactants[i]), *iter->get(),
```
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: should this be an unsigned int? this looks like it's casting to an int explicitly.

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp [hunk 5]
```diff
@@ -63,50 +77,52 @@ bool getReactantMatches(const MOL_SPTR_VECT &reactants,
...
+          matchesByReactant[i].push_back(match);
+        } else {
+          --matchCount;
         }
       }
-      if (keep) {
-        matchesByReactant[i].push_back(match);
-      } else {
-        --matchCount;
+      if (!matchCount) {
```
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: I'd rather this be written as matchCount > 0 to be more clear. Since matchCount is declared as an int above, I need mentally work out what happens if matchCount < 0

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp [hunk 6]
```diff
@@ -854,4 +882,192 @@ std::vector<MOL_SPTR_VECT> run_Reactants(const ChemicalReaction &rxn,
   return productMols;
 }  // end of ChemicalReaction::runReactants()
 
+// Generate a product based on a SINGLE reactant
```
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: a product or product(s)? this returns a vector of mols

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp [hunk 7]
```diff
@@ -854,4 +882,192 @@ std::vector<MOL_SPTR_VECT> run_Reactants(const ChemicalReaction &rxn,
...
+std::vector<MOL_SPTR_VECT> run_Reactant(const ChemicalReaction &rxn,
+                                        const ROMOL_SPTR &reactant,
+                                        unsigned int reactantIdx) {
```
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: Can we have a heuristic that figures out the reactantIdx automatically?  Maybe some heuristic such as using the MCS of the reactant with each of the reactant's rxns

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp [hunk 8]
```diff
@@ -854,4 +882,192 @@ std::vector<MOL_SPTR_VECT> run_Reactants(const ChemicalReaction &rxn,
...
+  Atom *rAtom;
+  Bond::BondType bond_type;
+  int mapno;
+  RGroup(Atom *atom, Bond::BondType type, int curmapno = -1)
+      : rAtom(atom), bond_type(type), mapno(curmapno) {}
+  RGroup(const RGroup &rhs)
+      : rAtom(rhs.rAtom), bond_type(rhs.bond_type), mapno(rhs.mapno) {}
+};
+}
+ROMol *reduceProductToSideChains(const ROMOL_SPTR &product, bool addDummyAtoms) {
```
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: You should probably mention in the docstring properties should be set on the product. Otherwise this functions looks fantastically magical from a first glance at the API :)

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp [hunk 9]
```diff
@@ -48,11 +48,25 @@ namespace RDKit {
 typedef std::vector<MatchVectType> VectMatchVectType;
 typedef std::vector<VectMatchVectType> VectVectMatchVectType;
 
+namespace {
```
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: how's the distinction made between what's inside common_properties and what's local SOME_UPPER_CASE_CONSTANT?

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp [hunk 10]
```diff
@@ -854,4 +882,192 @@ std::vector<MOL_SPTR_VECT> run_Reactants(const ChemicalReaction &rxn,
...
+      //  reaction mapping number to the scaffold_atom
+      //  sometimes this is a proper rgroup, so use that mapno
+      // C-C:12 >> C:12 # will probably work
+      //  C-C:12-C >>  C:12  # probably won't
+      int mapno = -1;
+      if (bonds_to_product.size()) {
+        mapno = getAtomMapNo(mol->getAtomBookmarks(), scaffold_atom);
+      }
+
+      atomsToRemove.push_back(
```
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: why can't this just be atomsToRemove.push_back(i)

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp [hunk 11]
```diff
@@ -854,4 +882,192 @@ std::vector<MOL_SPTR_VECT> run_Reactants(const ChemicalReaction &rxn,
...
+      ROMol::ADJ_ITER nbrIdx, endNbrs;
+      boost::tie(nbrIdx, endNbrs) = mol->getAtomNeighbors(scaffold_atom);
+      std::vector<RGroup> bonds_to_product;
+
+      while (nbrIdx != endNbrs) {
+        Atom *nbr = mol->getAtomWithIdx(*nbrIdx);
+        if (nbr->hasProp(REACT_ATOM_IDX)) {
+          if (nbr->hasProp(WAS_DUMMY)) {
+            bonds_to_product.push_back(RGroup(
+                nbr, mol->getBondBetweenAtoms(scaffold_atom->getIdx(), *nbrIdx)
```
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: scaffold_atom->getIdx() == i?

File: Code/GraphMol/ChemReactions/Reaction.h (hunk 3 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/Reaction.h]: Indeed, it is the same as runReactant, I'll just copy the verbiage.

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp (hunk 4 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: These aren't changes actually added by this pull request, the differences are indentation (it is now in an if statement).

Perhaps we should make a separate pull request to address these.  

I'm actually surprised we don't see this sign issue with -Wall -Weverything, we have been trying to pay attention to compiler warnings recently.

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp (hunk 6 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: It's worse, it is a vector of vector of mols, each reaction may have multiple product templates and each product template may have multiple "products".

It's kind of mind bending, nevertheless the product in the comment should be plural.

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp (hunk 7 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: That is certainly possible outside of the function, however it is still necessary to supply a reactantIdx as the reactant may actually match multiple reactant templates.

The intent is that this is really a utility function to help analyze reactions.  I.e. run a bunch of reagents through to see what side-chains they add.

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp (hunk 8 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: It is magical for now, and we should add that it only works on the output of runReactants or runReactant.

We are making a version that works regardless of the molecule, but is is a bit harder.

File: Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py [hunk 12]
```diff
@@ -83,7 +83,7 @@ def test1Basics(self):
       self.assertTrue(len(ps[0])==1)
       self.assertTrue(ps[0][0].GetNumAtoms()==3)    
 
-  def test2DaylightParser(self):
+  def atest2DaylightParser(self):
```
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py]: Crap, I pushed the wrong test code in (I turned off other tests when debugging)

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp (hunk 11 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: Yes, I'll replace i with scaffold_atom_idx to be more clear.

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp (hunk 9 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: There really is none, except we have the common_properties namespace to

1. protect typos ( I just realized that one of the BalabanJ entries is due to a typo from the original code that was automatically converted, there should only be one which kind of proves this point :) )
2. Eliminate the cost of converting a const char * into a std::string which was happening before the namespace was added.

These two const strings are not common, they are just used privately so they weren't added to the common_properties namespace, but they serve the same purpose as "typos" are caught by the compiler.

**Comment by bp-kelley** [comment]
@DoliathGavid Thanks for the thorough review.  Hopefully you'll see why this function was added when we start testing/pushing the new enumeration code.

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp (hunk 6 above)
Comment (DoliathGavid) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: Oh god. The SMARTS spec actually allows you to specify multiple product templates? I just read up on the spec, apparently this is how they implement logical OR for multiple reactant templates

[CH3:7][CH3:7] >> [CH3:7][CH3:8]

What a mess. Is this how rdkit implements multiple reactant/product templates as well? eg something like:


[CH3:8][CH3:7] >> [CH3:7][CH3:7]

**Comment by bp-kelley** [comment]
@greg fixed in @f7e5a2b15ab0955e24d9ee43606ec741920e1409

File: Code/GraphMol/ChemReactions/ReactionRunner.cpp (hunk 6 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/ReactionRunner.cpp]: Both smarts and the mdl format allow multiple reactions and products.  It actually makes sense with some reactions.

For smarts, it uses the same form for multiple products and reactants by using dot separated  strings on either side of ">>" as in this fake reaction below:

```
[C:1][N:2].[C:3]>>[C:1][N:2].[C:3]
```

Notice the "." separating the reactants and products.

**Comment by greglandrum** [comment]
Yep, there it is. Excellent. I'll go through this quickly one last time and then accept the PR.
